File: src/main/java/org/smarthata/rest/SensorController.java
package org.smarthata.rest;

import org.smarthata.model.Measure;
import org.smarthata.model.Sensor;
import org.smarthata.repository.MeasureRepository;
import org.smarthata.repository.SensorRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.web.SortDefault;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/sensors")
public class SensorController {

    private final SensorRepository sensorRepository;
    private final MeasureRepository measureRepository;

    @Autowired
    public SensorController(SensorRepository sensorRepository, MeasureRepository measureRepository) {
        this.sensorRepository = sensorRepository;
        this.measureRepository = measureRepository;
    }

    @GetMapping
    public Iterable<Sensor> findAll() {
        return sensorRepository.findAll();
    }

    @GetMapping("/{sensorId}")
    public Sensor findById(@PathVariable Integer sensorId) {
        return sensorRepository.findByIdOrElseThrow(sensorId);
    }

    @PutMapping("/{sensorId}")
    public Sensor put(@PathVariable Integer sensorId, @RequestBody Sensor sensor) {
        Sensor saved = sensorRepository.findByIdOrElseThrow(sensorId);
        saved.setName(sensor.getName());
        saved.setUnits(sensor.getUnits());
        return sensorRepository.save(saved);
    }

    @GetMapping("/{sensorId}/measures")
    public Page<Measure> measures(@PathVariable Integer sensorId,
                                  @SortDefault(value = "date", direction = Sort.Direction.DESC) Pageable pageable) {
        Sensor sensor = sensorRepository.findByIdOrElseThrow(sensorId);
        return measureRepository.findBySensor(sensor, pageable);
    }

}

File: src/main/kotlin/org/smarthata/rest/SensorController.kt
package org.smarthata.rest

import org.smarthata.model.Measure
import org.smarthata.model.Sensor
import org.smarthata.repository.MeasureRepository
import org.smarthata.repository.SensorRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.web.SortDefault
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/sensors")
class SensorController(
    private val sensorRepository: SensorRepository,
    private val measureRepository: MeasureRepository
) {
    @GetMapping
    fun findAll(): Iterable<Sensor> = sensorRepository.findAll()

    @GetMapping("/{sensorId}")
    fun findById(@PathVariable sensorId: Int?): Sensor = sensorRepository.findByIdOrElseThrow(sensorId)

    @PutMapping("/{sensorId}")
    fun put(@PathVariable sensorId: Int?, @RequestBody sensor: Sensor): Sensor {
        val saved = sensorRepository.findByIdOrElseThrow(sensorId)
        saved.name = sensor.name
        saved.units = sensor.units
        return sensorRepository.save(saved)
    }

    @GetMapping("/{sensorId}/measures")
    fun measures(
        @PathVariable sensorId: Int,
        @SortDefault(value = ["date"], direction = Sort.Direction.DESC) pageable: Pageable
    ): Page<Measure> {
        val sensor = sensorRepository.findByIdOrElseThrow(sensorId)
        return measureRepository.findBySensor(sensor, pageable)
    }
}